processing/src/main/java/io/druid/segment/DimensionSelectorUtils.java

```diff
@@ -170,8 +170,11 @@ private static ValueMatcher makeDictionaryEncodedValueMatcherGeneric(
       Predicate<String> predicate
   )
   {
-    final BitSet predicateMatchingValueIds = makePredicateMatchingSet(selector, predicate);
+    final BitSet checkedIds = new BitSet(selector.getValueCardinality());
```

Comment:
Better to use one bitset (with two-bit "blocks") than two separate bitsets, for locality. Also, in the future it could be replaced with specialized class that fetches and writes two bits at a time, in order to make less ops in the hot loop.

Reply:
I tried this transformation and it ended up slower. Timings after the change are below (compare to the numbers from the PR description). I'm not sure why. Is it possible that the JVM is so smart as to only compute things like bitIndex (inside the BitSets) once, in the case where there are two different BitSets and you get the same index from both in immediate succession?

```
Benchmark                                        (rowsPerSegment)  (schema)  Mode  Cnt    Score       Error  Units
FilterPartitionBenchmark.readWithExFnPostFilter            750000     basic  avgt   25  10690.447 ± 222.691  us/op
FilterPartitionBenchmark.readAndFilter                     750000     basic  avgt   25  305.222   ±   5.011  us/op
```

The transformed, slower code was:

```java
        @Override
        public ValueMatcher makeValueMatcher(final Predicate<String> predicate)
        {
          final BitSet matchingIds = new BitSet(getCardinality() * 2);

          // Lazy matcher; only check an id if matches() is called.
          return new ValueMatcher()
          {
            @Override
            public boolean matches()
            {
              final int id = getRowValue();
              final int checkBit = id * 2;
              final int matchBit = checkBit + 1;

              if (matchingIds.get(checkBit)) {
                return matchingIds.get(matchBit);
              } else {
                final boolean matches = predicate.apply(lookupName(id));
                matchingIds.set(checkBit);
                if (matches) {
                  matchingIds.set(matchBit);
                }
                return matches;
              }
            }

            @Override
            public void inspectRuntimeShape(RuntimeShapeInspector inspector)
            {
              inspector.visit("column", SimpleDictionaryEncodedColumn.this);
            }
          };
        }
        @Override
        public ValueMatcher makeValueMatcher(final Predicate<String> predicate)
        {
          final BitSet matchingIds = new BitSet(getCardinality() * 2);

          // Lazy matcher; only check an id if matches() is called.
          return new ValueMatcher()
          {
            @Override
            public boolean matches()
            {
              final int id = getRowValue();
              final int checkBit = id * 2;
              final int matchBit = checkBit + 1;

              if (matchingIds.get(checkBit)) {
                return matchingIds.get(matchBit);
              } else {
                final boolean matches = predicate.apply(lookupName(id));
                matchingIds.set(checkBit);
                if (matches) {
                  matchingIds.set(matchBit);
                }
                return matches;
              }
            }

            @Override
            public void inspectRuntimeShape(RuntimeShapeInspector inspector)
            {
              inspector.visit("column", SimpleDictionaryEncodedColumn.this);
            }
          };
        }
```

Reply:
Based on this experiment, and the other, simpler one below, I guess the JVM is that smart. So the original code is fastest, because it doesn't have as many branches and can also reuse computations internal to `BitSet.get`.

Reply:
> specialized class that fetches and writes two bits at a time, in order to make less ops in the hot loop.

This would probably be fastest of all, but since the current code is not too bad in terms of overhead even on the worse case, I think the optimization is not needed for this patch. I suspect that in the real world, it's more common to get the good case (the lazy optimization is useful and saves time) rather than the bad case (lazy optimization is useless, so we just hope it doesn't add too much overhead).

Reply:
> Is it possible that the JVM is so smart as to only compute things like bitIndex (inside the BitSets) once, in the case where there are two different BitSets and you get the same index from both in immediate succession?

Could you verify this theory by looking the assembly with `-prof perfasm` (`dtraceasm` on Mac)?

It's important to understand clearly why some counter intuitive results are observed.

Reply:
Hi @leventov do you have some tips on how to do this from within intellij on a Mac? I tried adding an annotation `@Fork(value = 1, jvmArgs = "-prof dtraceasm")` but I got a bunch of errors like this. I am hoping this is something you have run into in the past and may know how to fix.

```
Exception in thread "main" java.lang.NoClassDefFoundError: java/lang/invoke/LambdaForm$MH
	at com.sun.demo.jvmti.hprof.Tracker.nativeCallSite(Native Method)
	at com.sun.demo.jvmti.hprof.Tracker.CallSite(Tracker.java:99)
	at java.lang.invoke.InvokerBytecodeGenerator.emitNewArray(InvokerBytecodeGenerator.java:889)
	at java.lang.invoke.InvokerBytecodeGenerator.generateCustomizedCodeBytes(InvokerBytecodeGenerator.java:688)
	at java.lang.invoke.InvokerBytecodeGenerator.generateCustomizedCode(InvokerBytecodeGenerator.java:618)
	at java.lang.invoke.LambdaForm.compileToBytecode(LambdaForm.java:654)
	at java.lang.invoke.LambdaForm.prepare(LambdaForm.java:635)
	at java.lang.invoke.MethodHandle.<init>(MethodHandle.java:461)
	at java.lang.invoke.BoundMethodHandle.<init>(BoundMethodHandle.java:58)
	at java.lang.invoke.BoundMethodHandle$Species_L.<init>(BoundMethodHandle.java:211)
	at java.lang.invoke.BoundMethodHandle$Species_L.copyWith(BoundMethodHandle.java:228)
	at java.lang.invoke.MethodHandle.asCollector(MethodHandle.java:1002)
	at java.lang.invoke.MethodHandleImpl$AsVarargsCollector.<init>(MethodHandleImpl.java:460)
	at java.lang.invoke.MethodHandleImpl$AsVarargsCollector.<init>(MethodHandleImpl.java:454)
	at java.lang.invoke.MethodHandleImpl.makeVarargsCollector(MethodHandleImpl.java:445)
	at java.lang.invoke.MethodHandle.setVarargs(MethodHandle.java:1325)
	at java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(MethodHandles.java:1665)
	at java.lang.invoke.MethodHandles$Lookup.getDirectMethod(MethodHandles.java:1600)
	at java.lang.invoke.MethodHandles$Lookup.findStatic(MethodHandles.java:779)
	at java.lang.invoke.MethodHandleImpl$Lazy.<clinit>(MethodHandleImpl.java:627)
	at java.lang.invoke.MethodHandleImpl.varargsArray(MethodHandleImpl.java:1506)
	at java.lang.invoke.MethodHandleImpl.varargsArray(MethodHandleImpl.java:1623)
	at java.lang.invoke.MethodHandle.asCollector(MethodHandle.java:999)
	at java.lang.invoke.MethodHandleImpl$AsVarargsCollector.<init>(MethodHandleImpl.java:460)
	at java.lang.invoke.MethodHandleImpl$AsVarargsCollector.<init>(MethodHandleImpl.java:454)
	at java.lang.invoke.MethodHandleImpl.makeVarargsCollector(MethodHandleImpl.java:445)
	at java.lang.invoke.MethodHandle.setVarargs(MethodHandle.java:1325)
	at java.lang.invoke.MethodHandles$Lookup.getDirectMethodCommon(MethodHandles.java:1665)
	at java.lang.invoke.MethodHandles$Lookup.getDirectMethod(MethodHandles.java:1600)
	at java.lang.invoke.MethodHandles$Lookup.findStatic(MethodHandles.java:779)
	at java.lang.invoke.CallSite.<clinit>(CallSite.java:226)
	at java.lang.invoke.MethodHandleNatives.linkCallSiteImpl(MethodHandleNatives.java:307)
	at java.lang.invoke.MethodHandleNatives.linkCallSite(MethodHandleNatives.java:297)
	at java.io.ObjectInputStream.<clinit>(ObjectInputStream.java:3578)
	at org.openjdk.jmh.runner.link.BinaryLinkClient.<init>(BinaryLinkClient.java:74)
	at org.openjdk.jmh.runner.ForkedMain.main(ForkedMain.java:72)
HPROF ERROR: Unexpected Exception found afterward [hprof_util.c:494]
HPROF TERMINATED PROCESS
```

Reply:
I always run (and never heard anybody does anything different) JMH with profilers from the command line. Running benchmarks from an IDE is not a good idea anyway (might it affect the results that you have obtained previously?)

`-prof perfasm` is not a JVM arg, it's an arg of the JMH `Main`. Try `java -jar target/benchmarks -prof list` first.

If you are on Mac, you should use `dtraceasm`. You should also disable "System integrity protection" in order to be able to do this: http://osxdaily.com/2015/10/05/disable-rootless-system-integrity-protection-mac-os-x/

Reply:
I use a JMH IDEA plugin: https://github.com/artyushov/idea-jmh-plugin. Although I have never read the readme until now, and apparently it can affect the results slightly (they say they observed up to a couple of %).

What do you do to run the Druid JMH benchmarks? Do you know offhand what needs to be done to build them before they can be run? It'd be nice to add something to the Druid developer docs (https://github.com/druid-io/druid/blob/master/CONTRIBUTING.md) teaching people like me what to do.